File: bestia-zoneserver/src/test/java/net/bestia/zoneserver/chat/MapMoveCommandTest.java
package net.bestia.zoneserver.chat;

import net.bestia.entity.Entity;
import net.bestia.entity.EntityService;
import net.bestia.entity.component.PositionComponent;
import net.bestia.messages.MessageApi;
import net.bestia.messages.client.ToClientEnvelope;
import net.bestia.model.dao.AccountDAO;
import net.bestia.model.dao.MapParameterDAO;
import net.bestia.model.domain.Account;
import net.bestia.model.domain.MapParameter;
import net.bestia.model.geometry.Size;
import net.bestia.zoneserver.entity.PlayerEntityService;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Optional;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class MapMoveCommandTest {

	private final static long ACC_ID = 123;

	private MapMoveCommand cmd;

	@Mock
	private Account acc;

	@Mock
	private Entity entity;

	@Mock
	private AccountDAO accDao;

	@Mock
	private MessageApi akkaApi;

	@Mock
	private PlayerEntityService playerEntityService;

	@Mock
	private EntityService entityService;

	@Mock
	private MapParameterDAO mapParamDao;

	@Mock
	private MapParameter mapParam;

	@Mock
	private PositionComponent posComp;

	@Before
	public void setup() {

		when(acc.getId()).thenReturn(ACC_ID);

		when(mapParamDao.findFirstByOrderByIdDesc()).thenReturn(mapParam);

		when(mapParam.getWorldSize()).thenReturn(new Size(100, 100));

		when(entityService.getComponent(entity, PositionComponent.class)).thenReturn(Optional.of(posComp));

		when(playerEntityService.getActivePlayerEntity(anyLong())).thenReturn(entity);

		cmd = new MapMoveCommand(akkaApi, playerEntityService, entityService, mapParamDao);
	}

	@Test
	public void isCommand_okayCommand_true() {
		Assert.assertTrue(cmd.isCommand("/mm 10 11"));
		Assert.assertTrue(cmd.isCommand("/mm 10 10"));
	}

	@Test
	public void isCommand_falseCommand_false() {
		Assert.assertFalse(cmd.isCommand("/mmm"));
		Assert.assertFalse(cmd.isCommand("/mm2 34"));
		Assert.assertFalse(cmd.isCommand("/.mm 10 10"));
	}

	@Test
	public void executeCommand_wrongArgs_sendsMessage() {
		cmd.executeCommand(acc, "/mm bla bla");

		verify(akkaApi).send(any(ToClientEnvelope.class));
		verify(entityService, times(0)).updateComponent(posComp);
	}

	@Test
	public void executeCommand_validCords_setPositionToNewCords() {
		cmd.executeCommand(acc, "/mm 10 11");

		verify(posComp).setPosition(10, 11);
		verify(entityService).updateComponent(posComp);
	}

	@Test
	public void executeCommand_invalidCords_dontSetPosition() {

		cmd.executeCommand(acc, "/mm -10 11");
		verify(entityService, times(0)).updateComponent(posComp);
		verify(akkaApi).send(any(ToClientEnvelope.class));
	
		cmd.executeCommand(acc, "/mm 100000 11");

		verify(entityService, times(0)).updateComponent(posComp);
		verify(akkaApi, times(2)).send(any(ToClientEnvelope.class));
	}

	@Test
	public void executeCommand_entityWithNoPositionComp_doesNothing() {

		when(entityService.getComponent(entity, PositionComponent.class)).thenReturn(Optional.empty());

		cmd.executeCommand(acc, "/mm 10 11");

		verify(entityService, times(0)).updateComponent(posComp);
	}

}

File: bestia-zoneserver/src/test/kotlin/net/bestia/zoneserver/chat/MapMoveCommandTest.kt
package net.bestia.zoneserver.chat

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.check
import net.bestia.messages.client.ClientEnvelope
import net.bestia.model.account.Account
import net.bestia.zoneserver.actor.entity.EntityEnvelope
import net.bestia.zoneserver.actor.routing.MessageApi
import net.bestia.zoneserver.entity.PlayerEntityService
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class MapMoveCommandTest {
  private lateinit var cmd: MapMoveCommand

  @Mock
  private lateinit var acc: Account

  @Mock
  private lateinit var akkaApi: MessageApi

  @Mock
  private lateinit var playerEntityService: PlayerEntityService

  @BeforeEach
  fun setup() {
    cmd = MapMoveCommand(akkaApi, playerEntityService)
  }

  @Test
  fun isCommand_okayCommand_true() {
    assertTrue(cmd.isCommand("/mm 10 11"))
    assertTrue(cmd.isCommand("/mm 10 10"))
  }

  @Test
  fun isCommand_falseCommand_false() {
    assertFalse(cmd.isCommand("/mmm"))
    assertFalse(cmd.isCommand("/mm2 34"))
    assertFalse(cmd.isCommand("/.mm 10 10"))
  }

  @Test
  fun executeCommand_wrongArgs_sendsMessage() {
    cmd.executeCommand(acc, "/mm bla bla")

    verify(akkaApi).send(any<ClientEnvelope>())
  }

  @Test
  fun executeCommand_invalidCords_dontSetPosition() {
    cmd.executeCommand(acc, "/mm -10 11")
    verify(akkaApi).send(check<ClientEnvelope> {
      it.accountId == acc.id
    })

    cmd.executeCommand(acc, "/mm 100000 11")
    verify(akkaApi).send(any<EntityEnvelope>())
  }

  @Test
  fun executeCommand_entityWithNoPositionComp_doesNothing() {
    cmd.executeCommand(acc, "/mm 10 11")
  }

  companion object {
    private const val ACC_ID: Long = 123
  }
}
